Collection dependencies tab

jerabekjiri opened the pull request:
Issue: [AAH-752](https://issues.redhat.com/browse/AAH-752)

- Added displaying the number of dependencies in the `CollectionList` page. I had to add a plural option for the `NumericLabel` component. It was implemented just to put `s` at the end of the keyword, so what was happening => `123 Dependencys` and what we needed `2 Dependencies` (`2 Dependencys` => `2 Dependencies`).
![Screenshot from 2021-09-15 01-56-48](https://user-images.githubusercontent.com/19647757/133349131-ecb4e42d-27be-4d44-8723-e8507fc1558b.png)

### Dependencies page
- added `CollectionDependenciesList` component to display list of dependencies
- added `CollectionUsedbyDependenciesList` component to show all collections that are using this collection. 

![Screenshot from 2021-09-14 23-03-58](https://user-images.githubusercontent.com/19647757/133349207-97ecaf88-1ed5-4bea-a801-93ab4de9a2a7.png)

- Should we show an empty state on no `dependencies` or `usedby dependencies` list?
- Now, when I'm thinking about it, maybe `reverseDependencies` would be a better variable name than `usedByDependencies`? :smile:

To generate test collections run `make api/create-test-collections`. More here [#936](https://github.com/ansible/galaxy_ng/pull/936)

I still have some issues with generating test collections and therefore `Filter collection name` Input that uses `/collection-versions/?dependency=namespace.collection&name?=test_dependency` doesn't seem to be working on my end (pagination and sort works though), so I'll test this more when I figure it out.

jerabekjiri commented:
- fixed plural labels to apply to CollectionCard as well and reworked it to Constants
- added bulleted list to `CollectionDependenciesList`
- added empty state screens
  - Empty state when collection doesn't have dependencies and usedby dependencies 
![Screenshot from 2021-09-15 20-38-19](https://user-images.githubusercontent.com/19647757/133506994-3b4b73de-b329-4456-9c52-9f66c2db0245.png)
  - Empty state when no dependencies in collection
![Screenshot from 2021-09-15 21-32-03](https://user-images.githubusercontent.com/19647757/133507108-9ac456f5-fafc-45bc-b0ff-be1763a91adb.png)
 - Empty state when no usedby dependencies in collection 
![Screenshot from 2021-09-15 21-30-52](https://user-images.githubusercontent.com/19647757/133507201-85576fd2-1e12-4174-9f9b-f2637615696a.png)
  - Empty state on no results in usedby dependencies after filtering 
  
![Screenshot from 2021-09-15 22-39-43](https://user-images.githubusercontent.com/19647757/133507308-a8da1ff2-3cab-425d-a035-f953ae3dfe0b.png)

I'll ask Chris B. if the empty state texts are correct. And my collection filtering still doesn't work, need to look into it more :)

ZitaNemeckova commented on src/api/response-types/collection.ts:
```diff
@@ -11,6 +11,7 @@ export class CollectionVersion {
     contents: ContentSummaryType[];
     description: string;
     tags: string[];
+    dependencies: any; //FIXME: any for testing
```
Please create a dependency type instead of `any` :)

ZitaNemeckova commented on src/components/collection-dependencies-list/collection-usedby-dependencies-list.tsx:
```diff
@@ -0,0 +1,137 @@
+                    updateParams(ParamHelper.setParam(params, 'name', ''))
+                  }
+                  aria-label='filter-collection-name'
+                  placeholder={t`Filter collection name`}
```
Missing `by` and maybe `collection` is not needed -> "Filter by name".

jerabekjiri replied:
According to this [Dependencies prototype](https://marvelapp.com/prototype/a0c7h0j/screen/80780016/handoff) it's `Filter collection name`. Should I change it to `Filter by name` or ask Susan to verify?

ZitaNemeckova replied:
Let's verify with @sbuenafe-rh :)

ZitaNemeckova commented on src/containers/collection-detail/collection-dependencies.tsx:
```diff
@@ -0,0 +1,206 @@
+    return ParamHelper.updateParamsMixin();
+  }
+
+  private combineParams(...params) {
```
Unused function. Please remove.

jerabekjiri replied:
I'm using both functions in code, mostly like that
```
updateParams={(p) => {
   this.updateParams(this.combineParams(this.state.params, p), () =>
       this.loadData(true),
   );
}}
```

ZitaNemeckova replied:
My IDE lied to me :D Please ignore :)

ZitaNemeckova commented on src/containers/collection-detail/collection-dependencies.tsx:
```diff
@@ -0,0 +1,206 @@
+  }
+
+  private loadData(forceReload = false) {
+    this.loadCollection(forceReload, () => this.loadUsedByDependencies());
```
There should be a loading logic so a loading spinner aprears before data is loaaded instead of EmptyState.

ZitaNemeckova commented:
@jerabekjiri It seems to me that API endpoint doesn't filter/sort by name yet so I would drop the filter/sort and add it once the functionality is there. WDYT?

jerabekjiri commented:
@ZitaNemeckova thanks for the feedback, the filtering by name should be working after this [#974](https://github.com/ansible/galaxy_ng/pull/974)

jerabekjiri commented:
@ZitaNemeckova Fixed the issues, added the `LoadingPageSpinner` component for loading `usedByDependencies`, added the alert if an error occurs in loading dependencies request. Also PR#974 is merged now, so filtering by name should be working now.

cbudz commented:
Screens:

Empty state when no dependencies in collection

- replace "This collection is dependent on the following collections" with "This collections depends on the following collections" or even "This collections requires the following collections for use"

Empty state when no usedby dependencies in collection

- replace "No collection is using this dependency" with "Not required for use by other collections"

jerabekjiri commented:
Updated screen with changed messages 
![Screenshot from 2021-09-22 23-20-59](https://user-images.githubusercontent.com/19647757/134423870-f97ebabf-9c24-422f-aefb-e619e09bfa5a.png)

ZitaNemeckova reviewed (approved):
LGTM 👍 

Note: Need to ask API to do partial match.
